Re-implement Outgoing trait with Incoming associated type

DevinR528 opened the pull request.

mujx commented:
Maybe you need to run `ldconfig` after the libsodium compilation in the Dockerfile?

jimmycuadra commented:
That did the trick! Thanks!

DevinR528 committed:
Re-add the Outgoing trait and derive macro

The Outgoing trait now has no attributes except incoming_no_deserialize,
it looks for any references and lifetimes in a struct declaration and
removes them. The point of the Outgoing trait is to allow references to
be serialized and owned values to come out of deserialization. It has no
ability to wrap values in Raw (then EventResult).

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -187,17 +188,18 @@ impl ToTokens for Api {
-            } else {
-                TokenStream::new()
-            };
+        let typed_response_body_decl = if self.response.has_body_fields()
+            || self.response.newtype_body_field().is_some()
```
Do you understand why the formatting changed on these conditionals?

DevinR528 replied:
Hmm odd if I change it back `cargo fmt` just move it right back I wonder if an update to rustfmt at some point changed some metric?

jplatte replied:
🤷

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -187,17 +188,18 @@ impl ToTokens for Api {
+            || self.response.newtype_body_field().is_some()
+        {
+            quote! {
+                let response_body: <ResponseBody as ruma_api::Outgoing>::Incoming = ::ruma_api::try_deserialize!(
```
Since we're using proper absolute paths in these codeblocks otherwise, maybe to the same for `Outgoing`? (i.e. change `ruma_api` to `::ruma_api`)

jplatte commented on ruma-api-macros/src/derive_outgoing/wrap_incoming.rs:
```diff
@@ -0,0 +1,58 @@
+use syn::{
```
So we won't need this whole module and everything based on it anymore. We're still going to support `#[wrap_incoming]` without arguments and `#[incoming_no_deserialize]`, but that's it.

Let's talk on Matrix about the new transformations we need.

DevinR528 commented:
I could move the `ruma_api/tests/outgoing.rs` into the `trybuild` test section I left it to show what was compatible and it was helpful to have something to `cargo expand`.

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+        }
+        Data::Struct(s) => match s.fields {
+            Fields::Named(fs) => {
+                (fs.named.into_pairs().map(Pair::into_value).collect(), StructKind::Struct)
```
I think this is just `into_iter`:

```suggestion
                (fs.named.into_iter().collect(), StructKind::Struct)
```

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+        let mut found_wrap_incoming = false;
+
+        let mut remaining_attrs = Vec::new();
+        for attr in mem::replace(&mut field.attrs, Vec::new()) {
```
This can use `field.attrs.retain`, no?

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+            Path { leading_colon: None, segments }
+                if segments.len() == 1 && segments[0].ident == "incoming_no_deserialize" =>
+            {
+                return true
```
```suggestion
                return true;
```

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+    for attr in attrs {
+        match &attr.path {
+            Path { leading_colon: None, segments }
+                if segments.len() == 1 && segments[0].ident == "incoming_no_deserialize" =>
```
You probably just kept / copied old code here, right? Seems like this is just `is_ident("incoming_no_deserialize")`.

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+
+fn no_deserialize_in_attrs(attrs: &[Attribute]) -> bool {
+    for attr in attrs {
+        match &attr.path {
```
Can this be `if let Path { _ } && _`?

DevinR528 replied:
How's

```rust
for attr in attrs {
    if attr.path.is_ident("incoming_no_deserialize") {
        return true;
    }
}
```

jplatte replied:
Right, the matching is not required at all anymore. The above is definitely an improvement, but now that I see that I think the function can be further simplified to `attrs.iter().any(…)`.

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+
+fn split_for_impl_lifetime_less(generics: &mut Generics) -> (ImplGenerics, TypeGenerics) {
+    let mut remaining_params = Punctuated::<_, Token![,]>::new();
+    for param in mem::replace(&mut generics.params, Punctuated::<_, Token![,]>::new()) {
```
This seems like a long way to write

```rust
generics.params = generics.params.iter_mut().cloned().filter(/* sth. `matches!` */).collect();
```

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+                        }
+                    }
+
+                    *args = remaining_gen;
```
Same here, I think the equivalent iterator code would be more readable.

jplatte commented on ruma-api-macros/src/derive_outgoing.rs:
```diff
@@ -0,0 +1,184 @@
+                let TypePath { path, .. } = ty_path;
+                if path.is_ident("str") {
+                    field.ty = parse_quote! { String };
+                } else if path.is_ident("ServerName") || path.is_ident("DeviceId") {
```
This only works if `ServerName` or `DeviceId` is `use`d in the outer scope. We should probably support `ruma_identifiers::ServerName`, `ruma::identifiers::ServerName` and `ruma::ServerName` as well (plus the same for `DeviceId`, regardless of whether there is a leading `::`).

jplatte commented on ruma-api/tests/outgoing.rs:
```diff
@@ -0,0 +1,30 @@
+#[incoming_no_deserialize]
+pub struct Request<'a, T> {
+    pub abc: &'a str,
+
```
The empty lines in the test code seem pointless.

jplatte reviewed (approved):
Otherwise looks good.
